org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacCompilationUnitResolver.java

```diff
@@ -503,9 +502,30 @@ public CompilationUnit toCompilationUnit(org.eclipse.jdt.internal.compiler.env.I
 				.forEach(inMemoryCu -> {
 					pathToUnit.put(new String(inMemoryCu.getFileName()), inMemoryCu);
 				});
-
-		// note that this intentionally overwrites an existing working copy entry for the same file
-		pathToUnit.put(new String(sourceUnit.getFileName()), sourceUnit);
+		
+		if (!pathToUnit.keySet().isEmpty() && sourceUnit instanceof BasicCompilationUnit) {
+			// BasicCompilationUnit sometimes just has the last segment of the path,
+			// which means the working copy won't get overridden properly,
+			// resulting in duplicate classes being compiled.
+			List<String> realPaths = pathToUnit.keySet().stream().filter(path -> {
+				return path.endsWith(new String(sourceUnit.getFileName()));
+			}).toList();
+			if (!realPaths.isEmpty()) {
+				// HACK: if the user has multiple files with the same name opened,
+				// remove all of them then add the one under analysis back
+				if (realPaths.size() > 1) {
```

Comment:
Why not just overriding realPaths.get(0)` is it's the one we're replacing? Why do other units need to be discarded?

Reply:
OK, I think I got it. Can you please then just add to comment "remove all of them as one of them could be a duplicate of the one under analysis"?